File: services/sheets.py
import json
import os
from datetime import datetime, timezone
from functools import lru_cache

from google.oauth2 import service_account
from googleapiclient.discovery import build

_SPREADSHEET_ID = os.environ["GOOGLE_SPREADSHEET_ID"]
_SHEET_NAME = os.environ.get("GOOGLE_SHEET_NAME", "Gastos")
_SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]
_HEADERS = ["Fecha", "Monto", "Moneda", "Categoría", "Descripción", "Registrado en"]
_TZ = os.environ.get("TIMEZONE", "America/Argentina/Buenos_Aires")


@lru_cache(maxsize=1)
def _get_service():
    """Crea y cachea el cliente de Google Sheets (reutilizable entre invocaciones calientes)."""
    creds_json = os.environ.get("GOOGLE_CREDENTIALS_JSON")
    if creds_json:
        creds_info = json.loads(creds_json)
    else:
        creds_path = os.environ.get("GOOGLE_CREDENTIALS_PATH", "credentials.json")
        with open(creds_path) as f:
            creds_info = json.load(f)

    credentials = service_account.Credentials.from_service_account_info(
        creds_info, scopes=_SCOPES
    )
    return build("sheets", "v4", credentials=credentials, cache_discovery=False)


def _ensure_headers() -> None:
    service = _get_service()
    result = (
        service.spreadsheets()
        .values()
        .get(spreadsheetId=_SPREADSHEET_ID, range=f"{_SHEET_NAME}!A1:F1")
        .execute()
    )
    existing = (result.get("values") or [[]])[0]
    if not existing or existing[0] != _HEADERS[0]:
        service.spreadsheets().values().update(
            spreadsheetId=_SPREADSHEET_ID,
            range=f"{_SHEET_NAME}!A1:F1",
            valueInputOption="USER_ENTERED",
            body={"values": [_HEADERS]},
        ).execute()


def add_expense(expense: dict) -> None:
    _ensure_headers()
    service = _get_service()

    now_str = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
    row = [
        expense["date"],
        expense["amount"],
        expense["currency"],
        expense["category"],
        expense["description"],
        now_str,
    ]

    service.spreadsheets().values().append(
        spreadsheetId=_SPREADSHEET_ID,
        range=f"{_SHEET_NAME}!A:F",
        valueInputOption="USER_ENTERED",
        insertDataOption="INSERT_ROWS",
        body={"values": [row]},
    ).execute()


def get_recent_expenses(limit: int = 5) -> list[list]:
    """Retorna los últimos `limit` gastos (más reciente primero)."""
    service = _get_service()
    result = (
        service.spreadsheets()
        .values()
        .get(spreadsheetId=_SPREADSHEET_ID, range=f"{_SHEET_NAME}!A:F")
        .execute()
    )
    rows = (result.get("values") or [])[1:]  # Sin headers
    return list(reversed(rows[-limit:]))


def get_monthly_summary() -> dict:
    """Retorna totales del mes actual agrupados por moneda y categoría."""
    service = _get_service()
    result = (
        service.spreadsheets()
        .values()
        .get(spreadsheetId=_SPREADSHEET_ID, range=f"{_SHEET_NAME}!A:F")
        .execute()
    )
    rows = (result.get("values") or [])[1:]

    now = datetime.now()
    month_prefix = f"{now.year}-{now.month:02d}"

    month_rows = [r for r in rows if r and r[0].startswith(month_prefix)]

    summary: dict[str, dict] = {}
    for row in month_rows:
        try:
            amount = float(row[1])
            currency = row[2] if len(row) > 2 else os.environ.get("DEFAULT_CURRENCY", "ARS")
            category = row[3] if len(row) > 3 else "otros"
        except (ValueError, IndexError):
            continue

        if currency not in summary:
            summary[currency] = {"total": 0.0, "categories": {}}
        summary[currency]["total"] += amount
        summary[currency]["categories"][category] = (
            summary[currency]["categories"].get(category, 0.0) + amount
        )

    return {"summary": summary, "count": len(month_rows)}

File: src/services/sheets.js
'use strict';

const { google } = require('googleapis');
const path = require('path');
const config = require('../config');

let sheetsClient = null;

const HEADERS = ['Fecha', 'Monto', 'Moneda', 'Categoría', 'Descripción', 'Registrado en'];

/** Retorna (y cachea) el cliente autenticado de Google Sheets. */
async function getClient() {
  if (sheetsClient) return sheetsClient;

  let credentials;
  if (config.google.credentialsJson) {
    // Variable de entorno con JSON inline (Railway, Render, Heroku, etc.)
    credentials = JSON.parse(config.google.credentialsJson);
  } else {
    credentials = require(path.resolve(config.google.credentialsPath));
  }

  const auth = new google.auth.GoogleAuth({
    credentials,
    scopes: ['https://www.googleapis.com/auth/spreadsheets'],
  });

  const authClient = await auth.getClient();
  sheetsClient = google.sheets({ version: 'v4', auth: authClient });
  return sheetsClient;
}

/** Asegura que la fila de encabezados exista en la hoja. */
async function ensureHeaders() {
  const sheets = await getClient();
  const { spreadsheetId, sheetName } = config.google;

  const res = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range: `${sheetName}!A1:F1`,
  });

  const existingHeaders = res.data.values?.[0];
  if (!existingHeaders || existingHeaders[0] !== HEADERS[0]) {
    await sheets.spreadsheets.values.update({
      spreadsheetId,
      range: `${sheetName}!A1:F1`,
      valueInputOption: 'USER_ENTERED',
      requestBody: { values: [HEADERS] },
    });
  }
}

/**
 * Agrega una fila de gasto a la hoja.
 * @param {{ amount:number, currency:string, category:string, description:string, date:string }} expense
 */
async function addExpense(expense) {
  await ensureHeaders();

  const sheets = await getClient();
  const { spreadsheetId, sheetName } = config.google;

  const row = [
    expense.date,
    expense.amount,
    expense.currency,
    expense.category,
    expense.description,
    new Date().toLocaleString('es-AR', { timeZone: 'America/Argentina/Buenos_Aires' }),
  ];

  await sheets.spreadsheets.values.append({
    spreadsheetId,
    range: `${sheetName}!A:F`,
    valueInputOption: 'USER_ENTERED',
    insertDataOption: 'INSERT_ROWS',
    requestBody: { values: [row] },
  });
}

/**
 * Retorna los últimos N gastos de la hoja (más reciente primero).
 * @param {number} limit
 * @returns {Promise<string[][]>}
 */
async function getRecentExpenses(limit = 5) {
  const sheets = await getClient();
  const { spreadsheetId, sheetName } = config.google;

  const res = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range: `${sheetName}!A:F`,
  });

  const rows = (res.data.values || []).slice(1); // Sin headers
  return rows.slice(-limit).reverse();
}

/**
 * Calcula el resumen del mes actual agrupado por moneda y categoría.
 * @returns {Promise<{summary: Object, count: number}>}
 */
async function getMonthlySummary() {
  const sheets = await getClient();
  const { spreadsheetId, sheetName } = config.google;

  const res = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range: `${sheetName}!A:F`,
  });

  const rows = (res.data.values || []).slice(1);

  const now = new Date();
  const currentPrefix = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}`;

  const monthRows = rows.filter(r => r[0] && r[0].startsWith(currentPrefix));

  /** @type {Record<string, {total:number, categories:Record<string,number>}>} */
  const summary = {};

  for (const row of monthRows) {
    const amount = parseFloat(row[1]) || 0;
    const currency = row[2] || config.defaultCurrency;
    const category = row[3] || 'otros';

    if (!summary[currency]) summary[currency] = { total: 0, categories: {} };
    summary[currency].total += amount;
    summary[currency].categories[category] = (summary[currency].categories[category] || 0) + amount;
  }

  return { summary, count: monthRows.length };
}

module.exports = { addExpense, getRecentExpenses, getMonthlySummary };
